public/app/plugins/datasource/elasticsearch/components/QueryEditor/BucketAggregationsEditor/index.tsx

```diff
@@ -0,0 +1,37 @@
+import React, { FunctionComponent } from 'react';
+import { BucketAggregationEditor } from './BucketAggregationEditor';
+import { useDispatch } from '../../../hooks/useStatelessReducer';
+import { addBucketAggregation, removeBucketAggregation } from './state/actions';
+import { BucketAggregationAction } from './state/types';
+import { BucketAggregation } from './aggregations';
+import { useQuery } from '../ElasticsearchQueryContext';
+import { QueryEditorRow } from '../QueryEditorRow';
+import { IconButton } from '../../IconButton';
+
+interface Props {
+  nextId: BucketAggregation['id'];
+}
+
+export const BucketAggregationsEditor: FunctionComponent<Props> = ({ nextId }) => {
+  const dispatch = useDispatch<BucketAggregationAction>();
+  const { bucketAggs } = useQuery();
+  const totalBucketAggs = bucketAggs?.length || 0;
+
+  return (
+    <>
+      {bucketAggs!.map((bucketAgg, index) => (
```

Comment:
That's very true and I also thought about it, the reason I didn't focus on avoiding rerenderings due to `useQuery` is that (almost?) every time we use it we also use `useDispatch`, which uses ultimately depends on the `onChange` callback that is passed to the Query Editor.
The problem is that `onChange` also changes at every query change, which would basically nullify every optimization made by avoiding using `useQuery`.

> If so having it as a non optional prop would make it more obvious and you would not be able to use it without the bucketAggs

but I agree on this, it would probably be much cleaner. I'll give it a second pass.

Reply:
As I said no reason to do anything about it now in this PR, as you say there are probably other bottlenecks anyway and it is not a big issue.